src/torch_models.py

```diff
@@ -602,3 +606,85 @@ def classifier(self, state: State, objects: Sequence[Object]) -> bool:
         """
         v = state.vec(objects)
         return self._model.classify(v)
+
+
+################################## Utilities ##################################
+
+
+def _normalize_data(data: Array,
+                    scale_clip: float = 1) -> Tuple[Array, Array, Array]:
+    shift = np.min(data, axis=0)  # type: ignore
+    scale = np.max(data - shift, axis=0)  # type: ignore
+    scale = np.clip(scale, scale_clip, None)
+    return (data - shift) / scale, shift, scale
+
+
+def _balance_binary_classification_data(
+        X: Array, y: Array, rng: np.random.Generator) -> Tuple[Array, Array]:
+    pos_idxs_np = np.argwhere(np.array(y) == 1).squeeze()
+    neg_idxs_np = np.argwhere(np.array(y) == 0).squeeze()
+    pos_idxs = ([pos_idxs_np.item()]
+                if not pos_idxs_np.shape else list(pos_idxs_np))
+    neg_idxs = ([neg_idxs_np.item()]
+                if not neg_idxs_np.shape else list(neg_idxs_np))
+    assert len(pos_idxs) + len(neg_idxs) == len(y) == len(X)
+    keep_neg_idxs = list(
+        rng.choice(neg_idxs, replace=False, size=len(pos_idxs)))
+    keep_idxs = pos_idxs + keep_neg_idxs
+    X_lst = [X[i] for i in keep_idxs]
+    y_lst = [y[i] for i in keep_idxs]
+    X = np.array(X_lst)
+    y = np.array(y_lst)
+    return (X, y)
+
+
+def _train_predictive_pytorch_model(model: nn.Module,
+                                    loss_fn: Callable[[Tensor, Tensor],
+                                                      Tensor],
+                                    optimizer: optim.Optimizer,
+                                    tensor_X: Tensor,
+                                    tensor_Y: Tensor,
+                                    max_iters: int,
+                                    print_every: int = 1000,
+                                    clip_gradients: bool = False,
+                                    clip_value: float = 5,
+                                    n_iter_no_change: int = 10000000) -> None:
+    """Note that this currently does not use minibatches.
+
+    In the future, with very large datasets, we would want to switch to
+    minibatches.
+    """
+    model.train()
+    itr = 0
+    best_loss = float("inf")
+    best_itr = 0
+    model_name = tempfile.NamedTemporaryFile(delete=False).name
+    while True:
+        Y_hat = model(tensor_X)
+        loss = loss_fn(Y_hat, tensor_Y)
+        if loss.item() < best_loss:
+            best_loss = loss.item()
+            best_itr = itr
+            # Save this best model.
+            torch.save(model.state_dict(), model_name)
+        if itr % print_every == 0:
+            logging.info(f"Loss: {loss:.5f}, iter: {itr}/{max_iters}")
+        optimizer.zero_grad()
+        loss.backward()  # type: ignore
+        if clip_gradients:
```

Comment:
is this a new feature?

Reply:
no, we already had it for regressor training